velox/dwio/dwrf/writer/ColumnWriter.cpp

```diff
@@ -697,6 +697,88 @@ class TimestampColumnWriter : public BaseColumnWriter {
   std::unique_ptr<IntEncoder<false>> nanos_;
 };
 
+class DecimalColumnWriter : public BaseColumnWriter {
+ public:
+  DecimalColumnWriter(
+      WriterContext& context,
+      const TypeWithId& type,
+      const uint32_t sequence,
+      std::function<void(IndexBuilder&)> onRecordPosition)
+      : BaseColumnWriter{context, type, sequence, onRecordPosition},
+        type_{type.type()},
+        unscaledValues_{createDirectEncoder</* isSigned = */ true>(
+            newStream(StreamKind::StreamKind_DATA),
+            getConfig(Config::USE_VINTS),
+            type_->isShortDecimal() ? LONG_BYTE_SIZE : 2 * LONG_BYTE_SIZE)},
+        scales_{createRleEncoder</* isSigned = */ true>(
+            RleVersion_1,
+            // DWRF's NANO_DATA has the same enum value as ORC's SECONDARY.
+            newStream(StreamKind::StreamKind_NANO_DATA),
+            context.getConfig(Config::USE_VINTS),
+            LONG_BYTE_SIZE)} {
+    reset();
+  }
+
+  uint64_t write(const VectorPtr& slice, const common::Ranges& ranges)
+      override {
+    VELOX_CHECK(
+        slice->type()->equivalent(*type_),
+        "Unexpected vector type: {}.",
+        slice->type()->toString());
+    auto localDecoded = decode(slice, ranges);
+    auto& decodedVector = localDecoded.get();
+    writeNulls(decodedVector, ranges);
+
+    size_t count = 0;
+    auto [_, scale] = getDecimalPrecisionScale(*type_);
+
+    for (auto& pos : ranges) {
+      if (!decodedVector.mayHaveNulls() || !decodedVector.isNullAt(pos)) {
```

Comment:
Move the mayHaveNulls out of for loop

Reply:
Yes, we could also add fast path for flat-encoded vector. Will add these fast branches after https://github.com/facebookincubator/velox/pull/11431#discussion_r1832392079 is decided.

Reply:
Let's have one loop for may have nulls and one without? thanks!

Reply:
Thanks for the suggestion. Added separate paths according to 'mayHaveNulls' and the decimal type.